cmds/sshd/sshd.go

```diff
@@ -0,0 +1,238 @@
+			continue
+		}
+		channel, requests, err := newChannel.Accept()
+		if err != nil {
+			log.Fatalf("Could not accept channel: %v", err)
+		}
+
+		// Sessions have out-of-band requests such as "shell",
+		// "pty-req" and "env".  Here we handle only the
+		// "shell" request.
+		go func(in <-chan *ssh.Request) {
+			for req := range in {
+				dprintf("Request %v", req.Type)
+				switch req.Type {
+				case "shell":
+					if p == nil {
+						p, err = pty.New()
+						if err != nil {
+							log.Printf("sshd: pty.New failed, not running a shell")
+							break
+						}
+					}
+					err := runShell(channel, p, shell)
+					req.Reply(true, []byte(fmt.Sprintf("%v", err)))
+				case "pty-req":
+					p, err = newPTY(req.Payload)
+					req.Reply(err == nil, nil)
+				default:
+					fmt.Printf("Not handling req %v %q", req, string(req.Payload))
+					req.Reply(false, nil)
+				}
+			}
+		}(requests)
+
+	}
+}
+
+func main() {
+	flag.Parse()
+	if *debug {
+		dprintf = log.Printf
+	}
+	// Public key authentication is done by comparing
+	// the public key of a received connection
+	// with the entries in the authorized_keys file.
+	authorizedKeysBytes, err := ioutil.ReadFile(*keys)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	authorizedKeysMap := map[string]bool{}
+	for len(authorizedKeysBytes) > 0 {
+		pubKey, _, _, rest, err := ssh.ParseAuthorizedKey(authorizedKeysBytes)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		authorizedKeysMap[string(pubKey.Marshal())] = true
+		authorizedKeysBytes = rest
+	}
+
+	// An SSH server is represented by a ServerConfig, which holds
+	// certificate details and handles authentication of ServerConns.
+	config := &ssh.ServerConfig{
+		// Remove to disable public key auth.
+		PublicKeyCallback: func(c ssh.ConnMetadata, pubKey ssh.PublicKey) (*ssh.Permissions, error) {
+			if authorizedKeysMap[string(pubKey.Marshal())] {
+				return &ssh.Permissions{
+					// Record the public key used for authentication.
+					Extensions: map[string]string{
+						"pubkey-fp": ssh.FingerprintSHA256(pubKey),
+					},
+				}, nil
+			}
+			return nil, fmt.Errorf("unknown public key for %q", c.User())
+		},
+	}
+
+	privateBytes, err := ioutil.ReadFile(*privkey)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	private, err := ssh.ParsePrivateKey(privateBytes)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	config.AddHostKey(private)
+
+	// Once a ServerConfig has been configured, connections can be
+	// accepted.
+	listener, err := net.Listen("tcp", *ip+":"+*port)
+	if err != nil {
+		log.Fatal(err)
+	}
+	for {
+		nConn, err := listener.Accept()
+		if err != nil {
+			log.Fatalf("failed to accept incoming connection: %s", err)
+		}
+
+		// Before use, a handshake must be performed on the incoming
+		// net.Conn.
+		conn, chans, reqs, err := ssh.NewServerConn(nConn, config)
+		if err != nil {
+			log.Fatalf("failed to handshake: %v", err)
+		}
+		log.Printf("logged in with key %s", conn.Permissions.Extensions["pubkey-fp"])
```

Comment:
Could this log line include the remote IP? With the key signature and remote IP we'll have something that starting looking like a reasonable audit log.

Reply:
```
		log.Printf("%v logged in with key %s", conn.RemoteAddr(), conn.Permissions.Extensions["pubkey-fp"])
```